theseus/optimizer/nonlinear/trust_region.py

```diff
@@ -0,0 +1,182 @@
+        # with real reduction, we can increase the trust region
+        # in self.complete_step()
+        delta, self._at_trust_boundary_idx = self._compute_delta_impl()
+        if self._at_trust_boundary_idx is None:
+            self._at_trust_boundary_idx = (
+                TrustRegion._detached_squared_norm(delta) <= self._trust_region**2
+            )
+        return delta
+
+    @staticmethod
+    @torch.no_grad()
+    def _detached_squared_norm(
+        tensor: torch.Tensor, keepdim: bool = True
+    ) -> torch.Tensor:
+        return (tensor**2).sum(dim=1, keepdim=keepdim)
+
+    # Computes m_k(delta) as defined in Eq. (4.2) of the above reference (p. 68)
+    @torch.no_grad()
+    def _predicted_error(
+        self, previous_error: torch.Tensor, delta: torch.Tensor
+    ) -> torch.Tensor:
+        linearization = self.linear_solver.linearization
+
+        # TODO: For sparse I'll add a method like linearization.Avp() for Jacobian
+        # vector product. The issue is that for the sparse linearization we need
+        # to make this differentiable, because we are not using torch's
+        # native functions
+        assert isinstance(linearization, DenseLinearization)
+
+        # Note that B = |A^t@ A| so that p^tBp := delta^t A^t A delta = ||Adelta||^2
+        Adelta = linearization.A.bmm(delta.unsqueeze(2)).squeeze(2)
+        grad = -linearization.Atb.squeeze(2)
+        delta_dot_grad = (delta * grad).sum(dim=1)
+
+        return (
+            previous_error
+            + delta_dot_grad
+            + 0.5 * TrustRegion._detached_squared_norm(Adelta, keepdim=False)
+        )
+
+    @torch.no_grad()
+    def _complete_step(
+        self,
+        delta: torch.Tensor,
+        new_err: torch.Tensor,
+        previous_err: torch.Tensor,
+        accept_threshold: float = 0.0,
+        shrink_threshold: float = 0.25,
+        expand_threshold: float = 0.75,
+        shrink_ratio: float = 0.25,
+        expand_ratio: float = 2.0,
+        **kwargs,
+    ) -> Optional[torch.Tensor]:
+        # "err" tensors passed as input refer to the squared norm of the
+        # error vector, as returned by self.objective.error_squared_norm() / 2
+        good_params = (0.0 < shrink_ratio <= 1.0) and (expand_ratio >= 1.0)
+        good_params &= (shrink_threshold < expand_threshold) and (
+            accept_threshold < shrink_threshold
+        )
+        if not good_params:
+            raise ValueError(
+                "Invalid parameters for TrustRegionMethod. "
+                "Values must satisfy <accept/shrink>_threshold < expand_threshold, "
+                "shrink_ratio in (0, 1], and expand_ratio > 1.0."
+            )
+        pred_err = self._predicted_error(previous_err, delta)
+        rho = ((previous_err - new_err) / (previous_err - pred_err)).view(-1, 1)
+        shrink_idx = rho < shrink_threshold
+        self._trust_region = torch.where(
```

Comment:
For LM method, the trust-region should be scaled by the diagonals.

Reply:
While LM with adaptive damping is a trust region method, it doesn't inherit from `TrustRegion` because the current `_complete_step` logic is a bit different. I think it's def worth considering unifying these, but I'd prefer to leave for a separate PR to keep things simple.

Reply:
Ah, sorry. I think it might be better for the LM method to rescale the Hessian matrix as `H + damping_factor * diag(H)`.

Reply:
This is the `elliptical` damping option. We already have that.